### PR TITLE
Don't filter our partially fillable sell orders that have reached minBuyAmount

### PR DESCRIPTION
**Pull request by fleupold**
I'm not sure if this logic change is the correct fix, but the problem I'm seeing is that we currently filter out [this order](https://explorer.cow.fi/orders/0xdcb5921ee9cb5e3124e84c1c7ce0f24ec097ca86179704cfbebbc2c6eff00aa8fc78f8e1af80a3bf5a1783bb59ed2d1b10f78ca962f50b60) as being "in-flight" even though it is not. 

The reason is that in the current logic an order is filtered if either its buy or sell Amount are reached. This means a partially fillable order (e.g. sell 100 DAI for at least 50USDC) would be filtered if it was filled at a 50% at a price of 1:1 (50 DAI sold for 50 USDC). In this case we would have reached the sell amount and therefore no longer retain this order.

In this fix I changed the filled condition to be based on the order kind. A sell order is filled if its sell amount is reached, a buy order is filled when the buy amount is reached. Please review carefully as this is changing logic I'm not very familiar with (@vkgnosis I think you wrote the logic initially).

### Test Plan

Added two unit tests and remaining unit tests still pass.

### CODE REVIEW

File: crates/solver/src/in_flight_orders.rs [hunk 1]
```diff
@@ -70,11 +70,16 @@ impl InFlightOrders {
...
                 order.metadata.executed_buy_amount = u256_to_big_uint(&order.data.buy_amount);
+                order.metadata.executed_sell_amount = u256_to_big_uint(&order.data.sell_amount);
             }
         });
-        auction.orders.retain(|order| {
-            u256_to_big_uint(&order.data.buy_amount) > order.metadata.executed_buy_amount
-                && u256_to_big_uint(&order.data.sell_amount) > order.metadata.executed_sell_amount
+        auction.orders.retain(|order| match order.data.kind {
+            OrderKind::Sell => {
+                u256_to_big_uint(&order.data.sell_amount) > order.metadata.executed_sell_amount
```
Comment (nlordell) [inline crates/solver/src/in_flight_orders.rs:78]: I think it should be this:

```suggestion
                u256_to_big_uint(&order.data.sell_amount) > order.metadata.executed_sell_amount_before_fees
```

**Comment by nlordell** [comment]
> The reason is that in the current logic an order is filtered if either its buy or sell Amount are reached. This means a partially fillable order (e.g. sell 100 DAI for at least 50USDC) would be filtered if it was filled at a 50% at a price of 1:1 (50 DAI sold for 50 USDC)

This makes sense. And I think the code is better as is. What I don't understand is how we got to this case in the first place. Specifically, for partially fillable orders, I was under the impression that they were all liquidity orders, and should match at exactly their limit price.

**Comment by fleupold** [comment]
>What I don't understand is how we got to this case in the first place. Specifically, for partially fillable orders, I was under the impression that they were all liquidity orders, and should match at exactly their limit price.

image

**Comment by nlordell** [comment]
> Rounding

Ah OK - so its filtering out a tiny dust order.

**Review by nlordell — APPROVED** [review]

**Comment by nlordell** [comment]
Approved based on the last two commits - I assume this PR needs a rebase.

**Comment by vkgnosis** [comment]
Change looks correct to me. I don't think I wrote this code or if I did I don't remember it so I don't have special insight but the new code makes more sense.

**Review by vkgnosis — APPROVED** [review]

**Comment by fleupold** [comment]
> I don't think I wrote this code or if I did I don't remember it so I don't have special insight

Right, the original logic was added by @MartinquaXD in #152

**Review by MartinquaXD — APPROVED** [review]
Yeah, for some reason I forgot that `buy` and `sell` orders need to be treated differently in many situations. The new code seems correct now.

**Comment by fleupold** [comment]
Unit tests found another bug with using sell_amount_before_fee! 

I have to say it's quite annoying to have one giant test that is testing a bunch of orthogonal logic pieces. Makes it very hard to follow what exactly is breaking. I personally largely prefer to break the tests up into different scenarios.
